### ISSUE TITLE
Discovery of nodes on a single EC2 host

### ISSUE THREAD

**Issue by paulhoadley** [meltmedia/jgroups-aws#29]
At some point between releases 1.4.2 → 1.6.1, I seem to have lost the ability to discover JGroups stacks running on the _same host_ (in different JVMs). This coincided with the change from JGroups 3.6.8 → 4.0.5. My initial thought was misconfiguration, but I have tried a _lot_ of adjustments, and I'm not seeing it. To be clear, what I am describing is:

* JGroups running on separate hosts: `pbcast.GMS` reports first as co-ordinator, merges in second.
* JGroups running in different JVMs, different ports, same host: `pbcast.GMS` reports first as co-ordinator, then second as co-ordinator, and they don't see each other.

Has anyone seen this and/or can anyone offer any configuration advice? This is using configuration as supplied with this project.

**Comment by ctrimble** [comment]
@paulhoadley meltmedia is no longer using this project, so I have not seen any of the impacts of recent JGroups changes.  If the problem results in needing a new version of the project published, I am happy to support getting some updates released to Maven Central.

**Comment by paulhoadley** [comment]
Thanks @ctrimble. I am willing to put some time into debugging this, but the problem _I_ have is not really understanding how JGroups works. To help me get started: is it `AWS_PING` that's responsible for identifying JGroups nodes _on the same host,_ or some other part of the stack? To be clear, `AWS_PING` discovers nodes on _other_ EC2 instances just fine. I'm just not seeing it (or some other part of the stack) discovering nodes in different JVMs on the _same_ host.

**Comment by ctrimble** [comment]
I did a little digging in the commit history...

It looks like JGroups made some changes in the 3.6.x version around identifying cluster members.  Due to that, we picked up [this patch with logic that filters out the current machine](https://github.com/meltmedia/jgroups-aws/commit/0c2c083d54540cc73dc7ca02efa655b6308e68be#diff-0bcf1233d14ba946156c6d8e7cb805ea32bc1ab7ffe4e16b86430943587af405R339-R340).

That change was then refactored, but [the filtering remains](https://github.com/meltmedia/jgroups-aws/blob/develop/src/main/java/com/meltmedia/jgroups/aws/AWS_PING.java#L239).

That filter in the current code base just looks at the IP address.  Maybe we need to filter on host and port there, so that other cluster members on the same machine are visible.

**Comment by ctrimble** [comment]
@paulhoadley You might want to try changing the line:

```
       .filter(address -> !address.getIpAddress().equals(physical_addr.getIpAddress())) //filter out self
```

to something like:

```
        .filter(address -> address.compareTo(physical_addr)!=0) //filter out self
```

I assume that will get you going (read: I have not tested this).  It looks to me like the filtering currently assumes there is only one member per IP address.  The above change should take the port into account to, so that the other localhost members show up.

**Comment by paulhoadley** [comment]
Thanks again @ctrimble! I will definitely test this, hopefully in the next day or so.

**Comment by paulhoadley** [comment]
The single-line patch above worked: nodes on the _same host_ are now no longer discarded. _However,_ view stability now seems to be hosed. Fortunately, it can be pretty easily reproduced using `org.jgroups.demos.Chat` as described in this project's `README.md`. (Note that you probably want to bump the SLF-related dependencies in the POM up to at least 1.6.0, or it won't run due to not finding a logging implementation.) For reference, here's the output on two _different_ EC2 instances. First instance starts:

```
-------------------------------------------------------------------
GMS: address=ip-10-0-0-31-27549, cluster=ChatCluster, physical address=10.0.0.31:7800
-------------------------------------------------------------------
** view: [ip-10-0-0-31-27549|0] (1) [ip-10-0-0-31-27549]
```

then second:

```
-------------------------------------------------------------------
GMS: address=ip-10-0-0-237-37550, cluster=ChatCluster, physical address=10.0.0.237:7800
-------------------------------------------------------------------
** view: [ip-10-0-0-31-27549|1] (2) [ip-10-0-0-31-27549, ip-10-0-0-237-37550]
```

at which point first adds:

```
** view: [ip-10-0-0-31-27549|1] (2) [ip-10-0-0-31-27549, ip-10-0-0-237-37550]
```

And _that's it,_ it's then rock solid. In contrast, launching twice on the _same_ EC2 instance causes endless chatter of the form:

```
-------------------------------------------------------------------
GMS: address=ip-10-0-0-31-2548, cluster=ChatCluster, physical address=10.0.0.31:7800
-------------------------------------------------------------------
** view: [ip-10-0-0-31-2548|0] (1) [ip-10-0-0-31-2548]
> ** view: [ip-10-0-0-31-2548|1] (2) [ip-10-0-0-31-2548, ip-10-0-0-31-11255]
** view: [ip-10-0-0-31-2548|2] (1) [ip-10-0-0-31-2548]
** view: MergeView::[ip-10-0-0-31-11255|3] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-11255|2] (1) [ip-10-0-0-31-11255], [ip-10-0-0-31-2548|2] (1) [ip-10-0-0-31-2548]
** view: [ip-10-0-0-31-2548|4] (1) [ip-10-0-0-31-2548]
** view: MergeView::[ip-10-0-0-31-11255|5] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-2548|4] (1) [ip-10-0-0-31-2548], [ip-10-0-0-31-11255|4] (1) [ip-10-0-0-31-11255]
** view: [ip-10-0-0-31-2548|6] (1) [ip-10-0-0-31-2548]
** view: MergeView::[ip-10-0-0-31-11255|7] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-11255|6] (1) [ip-10-0-0-31-11255], [ip-10-0-0-31-2548|6] (1) [ip-10-0-0-31-2548]
** view: [ip-10-0-0-31-2548|8] (1) [ip-10-0-0-31-2548]
...
```

and on the second JVM:

```
-------------------------------------------------------------------
GMS: address=ip-10-0-0-31-11255, cluster=ChatCluster, physical address=10.0.0.31:7801
-------------------------------------------------------------------
** view: [ip-10-0-0-31-2548|1] (2) [ip-10-0-0-31-2548, ip-10-0-0-31-11255]
> ** view: [ip-10-0-0-31-11255|2] (1) [ip-10-0-0-31-11255]
** view: MergeView::[ip-10-0-0-31-11255|3] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-11255|2] (1) [ip-10-0-0-31-11255], [ip-10-0-0-31-2548|2] (1) [ip-10-0-0-31-2548]
** view: [ip-10-0-0-31-11255|4] (1) [ip-10-0-0-31-11255]
** view: MergeView::[ip-10-0-0-31-11255|5] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-2548|4] (1) [ip-10-0-0-31-2548], [ip-10-0-0-31-11255|4] (1) [ip-10-0-0-31-11255]
** view: [ip-10-0-0-31-11255|6] (1) [ip-10-0-0-31-11255]
** view: MergeView::[ip-10-0-0-31-11255|7] (2) [ip-10-0-0-31-11255, ip-10-0-0-31-2548], 2 subgroups: [ip-10-0-0-31-11255|6] (1) [ip-10-0-0-31-11255], [ip-10-0-0-31-2548|6] (1) [ip-10-0-0-31-2548]
...
```

You can see that there seems to be an initial consensus, but then indefinite re-negotiation of the cluster, including "subgroups".

Do you have any further thoughts, @ctrimble? I'm happy to keep testing this, but I'm at a bit of a loss as to how to fix it.

**Comment by ctrimble** [comment]
This makes me curious what would happen with standard PING on the same host outside EC2 with multiple nodes on the same machine.  Could this be a JGroups issue?

It also makes me wonder if compareTo is properly implemented for that address object, at least for how we are using it.  We could now be including ourself in the list.

**Comment by paulhoadley** [comment]
> This makes me curious what would happen with standard PING on the same host outside EC2 with multiple nodes on the same machine.

I presume we can't run both—can we? In my use case I need to support _both_ nodes on same host and nodes on other hosts. The curious bit, though, is that it definitely worked before. My current workaround is dropping back to `jgroups-aws` 1.4.2 (and JGroups 3), which has been working solidly in production for several years.

> Could this be a JGroups issue?

Not sure, but for what it's worth if you remove `-Dexec.args="-props conf/aws_ping.xml"` from the test command line (and hence run with vanilla JGroups), you see the same rock solid output as for two hosts above.

> It also makes me wonder if compareTo is properly implemented for that address object, at least for how we are using it. We could now be including ourself in the list.

Perhaps I can try to add some logging around that and find out. (Slightly difficult because it's in the middle of a stream, but I'll see what I can do.)

**Comment by ctrimble** [comment]
This is the code for [IpAddess compareTo](https://github.com/belaban/JGroups/blob/e41b66c519dd9af17c4c8fc771858fdfb5754fa6/src/org/jgroups/stack/IpAddress.java#L103).  It seems compatible with what we are doing.

**Comment by paulhoadley** [comment]
And now, _of course,_ I can no longer reproduce the issue. I've currently got three JVMs on the same host and one on another host, all talking to each other just fine:

```
** view: [ip-10-0-0-31-26069|5] (4) [ip-10-0-0-31-26069, ip-10-0-0-31-56372, ip-10-0-0-31-49988, ip-10-0-0-237-1343]
```

This makes _no sense_ to me. I didn't change anything. Just bad luck? Should we make your change to `findMembers()` and back away slowly? I might also bump JGroups up to 4.2.15.Final and report back. May as well update that while we're here.

**Comment by paulhoadley** [comment]
> I might also bump JGroups up to 4.2.15.Final and report back.

Looks solid. I simply cannot explain that weird merge frenzy above.

@ctrimble: give me a few hours and I'll bundle this up as a PR.

**Comment by paulhoadley** [comment]
I've submitted #30. If you think the JGroups 4.0.5 → 4.2.15 update is a bit much to bundle in with this fix, I'd be happy to back that out.

**Comment by paulhoadley** [comment]
I've had this running in a UAT environment for about three weeks now. Basic setup:

* Single EC2 host
* Two separate application instances/JVMs
* Light to moderate traffic
* Hundreds of messages logged passing between instances.

We saw a single error logged in the three week period:

```
ERROR org.jgroups.util.TimeScheduler3 - JGRP000169: failed executing task MERGE3: InfoSender
```

Underlying this was a 500 from AWS:

```
com.amazonaws.services.ec2.model.AmazonEC2Exception: An internal error has occurred (Service: AmazonEC2; Status Code: 500; Error Code: InternalError;
```

Both app instances logged this, and both recovered and continued to pass messages again for days.

In summary: this issue is fixed, and I will close this ticket.
